application/models/services/SurveyUpdater/LanguageConsistency.php

```diff
@@ -0,0 +1,137 @@
+     * @param LSYii_Application $yiiApp
+     */
+    public function __construct(
+        SurveyLanguageSetting $modelSurveyLanguageSetting,
+        LSYii_Application $yiiApp
+    ) {
+        $this->modelSurveyLanguageSetting = $modelSurveyLanguageSetting;
+        $this->yiiApp = $yiiApp;
+    }
+
+    /**
+     * Update
+     *
+     * @param Survey $survey
+     * @param string $initBaseLanguage
+     * @return void
+     */
+    public function update(Survey $survey, $initBaseLanguage)
+    {
+        // If the base language is changing, we may need the current title to avoid the survey
+        // being listed with an empty title.
+        $surveyTitle = $survey
+            ->languagesettings[$initBaseLanguage]
+            ->surveyls_title;
+
+        // Add new language fixLanguageConsistency do it
+        $aAvailableLanguages = $survey->getAllLanguages();
+        foreach ($aAvailableLanguages as $sLang) {
+            if ($sLang) {
+                $this->updateLanguage(
+                    $survey,
+                    $sLang,
+                    $surveyTitle
+                );
+            }
+        }
+
+        $this->cleanup(
+            $survey,
+            $initBaseLanguage,
+            $aAvailableLanguages
+        );
+    }
+
+    /**
+     * Update
+     *
+     * @param Survey $survey
+     * @param string $sLang
+     * @param string $surveyTitle
+     * @return void
+     */
+    private function updateLanguage(Survey $survey, $sLang, $surveyTitle)
+    {
+        $oLanguageSettings = $this->modelSurveyLanguageSetting->find(
+            'surveyls_survey_id=:surveyid AND surveyls_language=:langname',
+            array(':surveyid' => $survey->sid, ':langname' => $sLang)
+        );
+        if (!$oLanguageSettings) {
+            $oLanguageSettings = new SurveyLanguageSetting();
+            $languagedetails = getLanguageDetails($sLang);
+            $oLanguageSettings->surveyls_survey_id = $survey->sid;
+            $oLanguageSettings->surveyls_language = $sLang;
+            if ($sLang == $survey->language) {
+                $oLanguageSettings->surveyls_title = $surveyTitle;
+            } else {
+                $oLanguageSettings->surveyls_title = ''; // Not in default model ?
+            }
+            $oLanguageSettings->surveyls_dateformat = $languagedetails['dateformat'];
+            if (!$oLanguageSettings->save()) {
+                $this->yiiApp->setFlashMessage(
+                    gT("Survey language could not be created."),
+                    "error"
+                );
+                tracevar($oLanguageSettings->getErrors());
+            }
+        }
+    }
+
+    /**
+     * Update
+     *
+     * @param Survey $survey
+     * @param string $initBaseLanguage
+     * @param array $aAvailableLanguages
+     * @return void
+     */
+    private function cleanup(Survey $survey, $initBaseLanguage, $aAvailableLanguages)
+    {
+        fixLanguageConsistency(
```

Comment:
Bad how? Because its a global function?

Again this is copy and paste. This code is not currently triggered by any unit tests so I made no attempt to deal with it.

Reply:
Yeah I know. Can you write a todo, that fixLanguageConsistency should be lifted into this class at some point? Then just resolve this comment.

Reply:
There are a bunch on functions like this littered all over the code.

Reply:
Of course, but if we add TODOs in the service classes we can later grep for them for further cleanups. :)